### PR TITLE
Replace failure with thiserror

### PR DESCRIPTION
**Pull request by akhilles**
`failure` seems to be [deprecated](https://boats.gitlab.io/blog/post/failure-to-fehler/) and there are some compelling reasons to use `thiserror` instead:
- Better interoperability with other error wrapping solutions since `thiserror` doesn't introduce any custom traits
- It simply derives `std::error::Error` which means compatibility isn't broken if it's replaced down the road
- Better suited for libraries since it doesn't affect the public API and works well with [`anyhow`](https://docs.rs/anyhow/~1/anyhow/)

I would have opened an issue to discuss this, but I was already using the forked version so decided to just submit a PR.

### CODE REVIEW

File: libzmq/src/auth/server.rs [hunk 1]
```diff
@@ -46,10 +46,10 @@ impl fmt::Display for StatusCode {
     }
 }
 
-#[derive(Debug, Fail)]
-#[fail(display = "unable to parse status code")]
+#[derive(Debug, Error)]
+#[error("unable to parse status code")]
 #[doc(hidden)]
-pub struct StatusCodeParseError(());
```
Comment (jean-airoldie) [inline libzmq/src/auth/server.rs:52]: Whats the reasoning behind making this public?

File: libzmq/src/error.rs [hunk 2]
```diff
@@ -49,34 +40,36 @@ use std::{
 /// ```
 ///
 /// [`ErrorKind`]: enum.ErrorKind.html
-#[derive(Debug)]
-pub struct Error<T = ()> {
-    inner: Context<ErrorKind>,
+#[derive(Debug, Error)]
+#[error("{inner}")]
+pub struct Error<T: Debug = ()> {
```
Comment (jean-airoldie) [inline libzmq/src/error.rs:45]: Here adding a manual debug impl would be better so that the `T: Debug` bound is not needed.

File: libzmq/src/error.rs [hunk 3]
```diff
@@ -49,34 +40,36 @@ use std::{
...
+    #[source]
+    inner: ErrorKind,
     content: Option<T>,
```
Comment (jean-airoldie) [inline libzmq/src/error.rs:48]: Maybe this would be better in another PR but I think the error type should rather be:

```
pub struct Error<T = ()> {
    kind: ErrorKind,
    content: T
}
```

...and the `cast` method should be replaced by a `map` method akin to `Option::map`, so that `err.cast()` is replaced by `err.map(|_| ())`.

File: libzmq-sys/src/bindings.rs [hunk 4]
```diff
@@ -210,6 +210,9 @@ pub const ZMQ_CURRENT_EVENT_VERSION: u32 = 1;
 pub const ZMQ_CURRENT_EVENT_VERSION_DRAFT: u32 = 2;
 pub const ZMQ_EVENT_ALL_V1: u32 = 65535;
 pub const ZMQ_EVENT_ALL_V2: u32 = 131071;
+pub type __uint8_t = ::std::os::raw::c_uchar;
```
Comment (jean-airoldie) [inline libzmq-sys/src/bindings.rs:213]: What's up with that?

Comment (akhilles) [reply libzmq-sys/src/bindings.rs:213]: These were generated when doing a standard `cargo build`. I assume it's due to a bindgen update.

File: libzmq/src/auth/server.rs (hunk 1 above)
Comment (akhilles) [reply libzmq/src/auth/server.rs:52]: Assumed it wasn't part of public API, will fix

File: libzmq/src/error.rs (hunk 3 above)
Comment (akhilles) [reply libzmq/src/error.rs:46]: Makes sense, `get()` would probably be:
```rust
fn get(&self) -> &T
```

What about `take()`?
```rust
fn take(self) -> T
```
or just
```rust
impl<T> Into<T> for Error<T>
```

Comment (akhilles) [reply libzmq/src/error.rs:46]: or maybe:
```rust
fn take(self) -> (Error<()>, T)
```

Comment (jean-airoldie) [reply libzmq/src/error.rs:46]: I think a `fn into_inner(self) -> T` might be more obvious than a `Into<T>` impl.

File: libzmq/src/error.rs [hunk 5]
```diff
@@ -104,30 +95,23 @@ impl<T> Error<T> {
     ///
     /// This is not implemented as `Into<Error<I>>` to be explicit since
     /// information is lost in the conversion.
-    pub fn cast<I>(self) -> Error<I> {
+    pub fn cast<I: fmt::Debug>(self) -> Error<I> {
```
Comment (jean-airoldie) [inline libzmq/src/error.rs:98]: That bound is no longer needed right?

**Comment by jean-airoldie** [comment]
I think that's its fine if the `Debug` implementation of `Error` does not output the `content`. So you could add a separate manual implementation such as:
```rust
impl fmt::Debug for Error {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.debug_struct("Error")
           .field("kind", &self.kind)
           .finish()
    }
}
```

This would allow the `T: Debug` bound to be removed everywhere.

File: libzmq/src/error.rs [hunk 6]
```diff
@@ -178,24 +167,19 @@ impl<T> From<Error<T>> for io::Error {
...
-/// # Note
-/// This error type is non-exhaustive and could have additional variants
-/// added in future. Therefore, when matching against variants of
-/// non-exhaustive enums, an extra wildcard arm must be added to account
-/// for any future variants.
-///
 /// [`Error`]: enum.Error.html
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Fail, Hash)]
+#[derive(Debug, Error, Copy, Clone, PartialEq, Eq, Hash)]
+#[non_exhaustive]
```
Comment (akhilles) [inline libzmq/src/error.rs:172]: `#[non_exhaustive]` bumps MSRV to 1.40. Is this fine?

Comment (jean-airoldie) [reply libzmq/src/error.rs:172]: Sure.

**Comment by jean-airoldie** [comment]
Looks good. Can I merge?

**Comment by akhilles** [comment]
> Looks good. Can I merge?

Yup!

**Comment by jean-airoldie** [comment]
Good stuff!
